src/RestApi/StoreApi/Routes/Checkout.php

```diff
@@ -357,7 +357,16 @@ protected function process_payment( \WC_Order $order, \WP_REST_Request $request
 	 * @return string
 	 */
 	protected function get_request_payment_method_id( \WP_REST_Request $request ) {
-		$payment_method = wc_clean( wp_unslash( $request['payment_method'] ) );
+		$payment_data = $this->get_request_payment_data( $request );
```

Comment:
The issue I'm working around here is that in the client the `active_payment_method` is set to a unique id across all payment methods (including express).  So we can't have both `stripe` as the id for the stripe cc payment method and `stripe` as the id in the client for the payment request payment method. Instead the payment methods are registered as `stripe` for stripe cc and `payment_request` for express payment.  This could be further compounded when the blocks get support for `stripe SEPA` etc payment methods.

So, we either fix this client side or server side.  I chose server side for now because it's trivial for payment methods to include what the payment method id is in the payment_data package. 

I'm open to other suggestions here :)

Reply:
That's what I'd expect; stripe cc would have ID `stripe`, the express one would have something like `stripe_apple_pay`, and sepa would be `stripe_sepa` -- but what I don't see is why on the server side, these unique IDs cannot be handled together? Maybe Im missing a piece of the puzzle.

Reply:
Some additional info.  Currently in WooCommerce core (and payment extensions) server side:

- class-wc-gateway-stripe is used for stripe cc and stripe payment request, but stripe payment request adds details on the request type (apple-pay, chrome pay etc) via order-meta before passing off the checkout for processing.
- payment request payment method actually bypasses the checkout form submission and does it's own submission via ajax.
- Other stripe payment methods have their own gateway files (so you're correct there, having the ids just be `stripe_sepa` etc will map fine for dynamic gateway loading.

So the primary issue here is that the stripe express payment methods are a "special class" of payment method that still use the `class-wc-gateway-stripe` for the main payment processing but also add additional order meta information.

So the active payment id for both stripe cc and stripe payment request is `stripe`.  

I also think I implemented this initially because with express payment methods the "active payment method" in the checkout form might differ from the payment-method-id used for an express payment so this ensures that checkout processing is definitely using the correct id.

Even as i write this, I know the logic probably can be reworked.  I just don't have a good idea of what to rework it to yet. Maybe, if the express payment modal is closed because it's payment was successful we don't restore active payment method to the checkout payment method? Maybe we always have `payment_method` be a property of `payment_data` instead of a top-level property in the request (so express payment methods can just overwrite it) and detached from `activePaymentMethod` in the checkout state (and that prop is simply used for indicating what payment method is active in the embedded payment method options)?

Reply:
> So the primary issue here is that the stripe express payment methods are a "special class" of payment method that still use the class-wc-gateway-stripe for the main payment processing but also add additional order meta information.

The confusing thing for me is why then this needs to be munged at all. If the method is `stripe`, `get_request_payment_method_id` is going to return that and validation will pass because `stripe` is valid. 

`woocommerce_rest_checkout_process_payment_with_context` is going to give you both the ID (`stripe`) and the `payment_data` (which may contain this sub-payment method ID and other data) - so why isn't that enough to handle the request and/or add additional data?

Reply:
> The confusing thing for me is why then this needs to be munged at all. If the method is stripe, get_request_payment_method_id is going to return that and validation will pass because stripe is valid.

it's because the payment_id coming in through the api is `payment_request` (being `$request['payment_method']` is set from `activePaymentMethod` payment side. So it's not `stripe`.

Reply:
To be clear, I'm in agreement that it'd be better to avoid this logic server side.  Would you be okay with me adding this as a follow-up to improve or are you wanting to handle this now?

Reply:
@nerrad Depends how hard to resolve and for when; if it's a follow up we should definitely tackle before making the API public because I wouldn't want this to ship with the final version.

> t's because the payment_id coming in through the api is payment_request (being $request['payment_method'] is set from activePaymentMethod payment side. So it's not stripe.

I see. So is it best to let the payment method change that client side before firing the request, or allow some mechanism to swap it server side? On the server side we could add some filter within `get_request_payment_method_id` that lets you change it e.g. `woocommerce_rest_checkout_request_payment_method_id`, or we could lax the validation so that any ID gets passed to `woocommerce_rest_checkout_process_payment_with_context` and if unhandled, then it throws the error.

Reply:
Having `woocommerce_rest_checkout_process_payment_with_context` run even if the method is not recognised might be the best way forward here as it requires less code on the payment method side. We would just want to provide a better error response if the `woocommerce_rest_checkout_process_payment_with_context` action yields no results.

Reply:
>  if it's a follow up we should definitely tackle before making the API public because I wouldn't want this to ship with the final version.

I agree.  I just don't want the headache of rebasing on this pull and assuming I can get tests to pass, merging enables express checkout for testing and further work.

Reply:
> We would just want to provide a better error response if the woocommerce_rest_checkout_process_payment_with_context action yields no results

So would this be a matter of checking that `PaymentResult` reference has the expected internal state?

Reply:
Possibly yes; detect if the result's status is the default so we know it wasn't touched.

What does come through as method ID though? Does it matter that it's saved to the order object? Do we need to adjust that too?

Reply:
Method ID is used to determine what gateway class to load for payment processing.  `payment_request_type` is specific to stripe (well at least what I've seen so far) and is used to record what specific payment_request api type was used to make the payment (apple pay, chrome pay etc) which is then displayed (I'm assuming by stripe gateway extension hooking into Woo) in various order details.